mvpoly/src/lib.rs

```diff
@@ -261,3 +261,62 @@ pub trait MVPoly<F: PrimeField, const N: usize, const D: usize>:
     /// variable in each monomial is of maximum degree 1.
     fn is_multilinear(&self) -> bool;
 }
+
+/// Compute the cross terms of a list of polynomials. The polynomials are
+/// linearly combined using the power of a combiner, often called `α`.
+/// Powers of the combiner are considered as an additional variable of the
+/// multi-variate polynomials, therefore increasing the degree of the sum of all
+/// polynomials by one. In other words, if we combine (M + 1) polynomials
+/// `P_1(X_1, ..., X_N)`, ..., `P_{M + 1}(X_1, ..., X_N)`, we will compute the
+/// cross-terms of the polynomial:
+///
+/// ```text
+/// P(X_1, ..., X_N, α_1, ..., α_M) =     P_1(X_1, ..., X_N)
+///                                 + α_1 P_2(X_1, ..., X_N)
+///                                 + ...
+///                                 + α_M P_{M + 1}(X_1, ..., X_N)
+/// ```
+///
+/// We notice that we simply do have a shifted linear combination of the
+/// polynomials. Based on the property that the cross-terms of a sum of
+/// polynomials is the sum of the cross-terms of the individual polynomials, and
+/// that shifting a polynomial with a new variable simply requires to shift the
+/// power of cross-terms by one and multiply by the evaluation at the new point,
+/// we get a simple algorithm.
+///
+/// All polynomials are supposed to be given as multivariate polynomials of the
+/// same degree and the same number of variables, even if they do not have the
+/// same number of variables nor the same degree.
+/// Therefore, `N` is the maximum number of variables of the polynomials and `D - 1`
+/// is the maximum degree of the polynomials, `D` being the the degree of the
+/// combined polynomial when including the combiner.
+pub fn compute_combined_cross_terms<
+    F: PrimeField,
+    const N: usize,
+    const D: usize,
+    T: MVPoly<F, N, D>,
+>(
+    polys: Vec<T>,
+    combiner1: F,
+    combiner2: F,
+    eval1: [F; N],
+    eval2: [F; N],
+    u1: F,
+    u2: F,
```

Comment:
I m not sure I understand what is computed here ?
the text mentioned combining the polys with one combiner, and we have two of them here.
I understand that the two combiners are the alpha from each instances we fold ? 
But then I don't get the assymetrical role ?

Reply:
See https://hackmd.io/@dannywillems/Syo5MBq90#Computation-of-cross-terms-for-shifted-polynomials